sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/CharsetDetector.java

```diff
@@ -0,0 +1,132 @@
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.charset.CharacterCodingException;
+import java.nio.charset.Charset;
+import java.nio.charset.CharsetDecoder;
+import java.nio.charset.CodingErrorAction;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+import javax.annotation.CheckForNull;
+
+import org.apache.commons.io.ByteOrderMark;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.io.input.BOMInputStream;
+
+public class CharsetDetector {
+  private static final int BYTES_TO_DECODE = 512;
+  private Path filePath;
+  private BOMInputStream stream;
+  private Charset detectedCharset;
+  private Charset defaultEncoding;
+
+  public CharsetDetector(Path filePath, Charset defaultEncoding) {
+    this.filePath = filePath;
+    this.defaultEncoding = defaultEncoding;
+  }
+
+  public boolean run() {
+    try {
+      stream = createInputStream(filePath);
+      if (detectBOM()) {
+        return true;
+      }
+
+      return detectCharset();
+    } catch (IOException e) {
+      throw new IllegalStateException("Unable to read file " + filePath.toAbsolutePath().toString(), e);
+    }
+  }
+
+  @CheckForNull
+  public Charset charset() {
+    assertRun();
+    return detectedCharset;
+  }
+
+  public InputStream inputStream() {
+    assertRun();
+    return stream;
+  }
+
+  private static BOMInputStream createInputStream(Path path) throws IOException {
+    BufferedInputStream bufferedStream = new BufferedInputStream(Files.newInputStream(path));
+    return new BOMInputStream(bufferedStream, ByteOrderMark.UTF_8, ByteOrderMark.UTF_16LE,
+      ByteOrderMark.UTF_16BE, ByteOrderMark.UTF_32LE, ByteOrderMark.UTF_32BE);
+  }
+
+  private boolean detectBOM() throws IOException {
+    String charsetName = stream.getBOMCharsetName();
+    if (charsetName != null) {
+      detectedCharset = Charset.forName(charsetName);
+      return true;
+    }
+    return false;
+  }
+
+  @CheckForNull
+  private boolean detectCharset() throws IOException {
+    stream.mark(BYTES_TO_DECODE);
+    byte[] buf = new byte[BYTES_TO_DECODE];
+    int len = IOUtils.read(stream, buf, 0, BYTES_TO_DECODE);
+    stream.reset();
+
+    Set<Charset> charsets = new LinkedHashSet<>();
+    charsets.add(defaultEncoding);
+    charsets.add(StandardCharsets.UTF_8);
+    charsets.add(Charset.defaultCharset());
+
+    for (Charset c : charsets) {
+      if (tryDecode(buf, len, c)) {
+        detectedCharset = c;
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private static boolean tryDecode(byte[] bytes, int len, Charset charset) throws IOException {
+    CharsetDecoder decoder = charset.newDecoder()
+      .onMalformedInput(CodingErrorAction.REPORT)
+      .onUnmappableCharacter(CodingErrorAction.REPORT);
+
+    try {
+      decoder.decode(ByteBuffer.wrap(bytes, 0, len));
+    } catch (CharacterCodingException e) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1166)